### PR TITLE
pFaTools

### PR DESCRIPTION
**Pull request by saqzhao**
# Description

Added command line usage for pFasta tools: Sample, Extract, ExtractBed.
- ```extract in.pfa chrom start end out.pFa (optional chrom)```: start and end are integer indices indicating the desired subsequence, and the user can provide an optional string chrom to specify which sequence in in.pfa it is a subsequence of. This subsequence is written as a new pFasta out.pFa.
- ```extractBed in.pfa regions.bed outDir```: regions.bed specifies the chrom, start and end for the desired subsequence from in.pfa and writes a new pFasta in outDir.
- ```sample in.pfa chrom outDir``` takes an input pFasta and target sequence name. It writes to outdir a Fasta sampled from the distribution specified by the input and chrom.

- Added comment about usage to AreEqual in fileio/fileio.go.

### Testing

Tests exist for Sample, Extract and ExtractBed (using the same input and expected output as in fasta/pfasta/tools_test.go)

### Checklist before requesting a review

- [x] I performed a self-review of my code
- [x] If it this a core feature, I added thorough tests
- [x] The command `go fmt` or `make clean` was used on all files included

### CODE REVIEW

File: fasta/pFasta/tools_test.go [hunk 1]
```diff
@@ -286,39 +280,58 @@ var SampleTests = []struct {
...
+		bed.Write(InRegion, bedInput)
+
 		res := ExtractBed(testCase.Input, testCase.Region, testCase.OutputName)
-		fmt.Print(res)
+
 		if !IsEqual(res, testCase.Expected, testCase.Precision) {
 			t.Errorf("Error: in pFasta. ExtractBed valid input test not as expected.\n")
 		}
+
+		OutFile := "testdata_tools/test_extractbed_expected.pfa"
```
Comment (rimangan) [inline fasta/pFasta/tools_test.go:320]: I think these lines can be removed now

**Review by rimangan — COMMENTED** [review]
Great work! A few comments on my end

**Comment by cfauci** [comment]
Thanks for the PR contribution! Can you provide a summary of the purpose of the tools in this PR so we know what they will do before we start reviewing and then bulk up the documentation? I wasn't sure what the function descriptions were saying, or what the usage statements for the subcommands meant.

File: cmd/pFaTools/extract.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		Start:   parse.StringToInt(start),
+		End:     parse.StringToInt(end),
+	}
+
+	pFaExtract(s)
+}
+
+// pFaExtract parses an input pFasta file and extracts the file according to user-defined settings.
+func pFaExtract(s ExtractSettings) {
+	records := []pFasta.PFasta{pFasta.Extract(pFasta.Read(s.InFile)[0], s.Start, s.End, s.OutName)}
```
[the remark this answers is not in the document]
Comment (saqzhao) [reply cmd/pFaTools/extract.go:70]: Resolved, added chrom as a parameter for pfasta.extract

File: cmd/pFaTools/extract.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	End     int
+}
+
+// extractUsage defines the usage statement for the pFaTools extract subcommand.
+func extractUsage(extractFlags *flag.FlagSet) {
+	fmt.Printf("pFaTools extract - a tool for extracting a subsequence from a pFa.\n" +
+		"A note on input and output file types:\n" +
+		"\tpFa (probabilistic FASTA) encodes position-wise probabilities in matrix entries. Columns should therefore sum to 1.\n" +
+		"Usage:\n" +
+		"pFaTools extract in.pFa chrom out.fasta start end\n" +
```
Comment (rimangan) [inline cmd/pFaTools/extract.go:29]: Output of extract is a .pfa, not a .fasta

File: cmd/pFaTools/extractBed.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+// extractBedUsage defines the usage statement for the pFaTools extractBed subcommand.
+func extractBedUsage(extractBedFlags *flag.FlagSet) {
+	fmt.Printf("pFaTools extractBed - a tool for extracting a subsequence from a pFa, as specified by a Bed file.\n" +
+		"A note on input and output file types:\n" +
+		"\tpFa (probabilistic FASTA) encodes position-wise probabilities in matrix entries. Columns should therefore sum to 1.\n" +
+		"\tbed (browser extensible data) stores genomic regions as coordinates and associated annotations." +
+		"Usage:\n" +
+		"pFaTools extractBed in.pFa region.bed outFile\n" +
```
Comment (rimangan) [inline cmd/pFaTools/extractBed.go:28]: Can we add an extension to outfile, in this case outFile.pFa

File: cmd/pFaTools/extractBed.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+type ExtractBedSettings struct {
+	InFile  string
+	Region  string
+	OutFile string
+	TakeCoords bool
+}
+
+// extractBedUsage defines the usage statement for the pFaTools extractBed subcommand.
+func extractBedUsage(extractBedFlags *flag.FlagSet) {
+	fmt.Printf("pFaTools extractBed - a tool for extracting a subsequence from a pFa, as specified by a Bed file.\n" +
```
Comment (rimangan) [inline cmd/pFaTools/extractBed.go:23]: Can we update this usage to reflect the fact that now we can extract multiple regions at once?

File: cmd/pFaTools/extractBed.go (hunk 4 above)
Comment (saqzhao) [reply cmd/pFaTools/extractBed.go:28]: Added!

File: cmd/pFaTools/extractBed.go (hunk 5 above)
Comment (saqzhao) [reply cmd/pFaTools/extractBed.go:23]: Added!

File: cmd/pFaTools/extract.go (hunk 3 above)
Comment (saqzhao) [reply cmd/pFaTools/extract.go:29]: Changed

**Review by rimangan — APPROVED** [review]

**Review by cfauci — APPROVED** [review]

File: fasta/pFasta/tools.go [hunk 6]
```diff
@@ -2,65 +2,88 @@ package pFasta
...
+	var outName string
+	if takeCoords {
+		outName = fmt.Sprintf("%s:%v-%v", chrom, start, end)
+	} else if len(outputName) > 0 {
+		outName = outputName
+	} else {
+		outName = chrom
+	}
+
+	var answer = PFasta{Name: outName, Seq: make([]pDna.Float32Base, end-start)}
```
Comment (jayashabarishankar) [inline fasta/pFasta/tools.go:54]: Why is var answer defined twice here?

**Pull request merged**